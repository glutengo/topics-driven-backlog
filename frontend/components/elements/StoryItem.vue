<template>
    <tr :class="{ [$style.highlight]: highlight || sortable && keyboardSorting, 'story-item': true }">
        <td 
            v-if="editing" 
            colspan="6">

            <StoryForm
                :project="project"
                v-model="editingData" 
                @cancel="cancelEdit"
                @submit="save"/>
        </td>

        <template v-else>
            <!-- Drag and drop -->
            <td v-if="isView(['backlog', 'planning-sprint', 'planning-backlog'])" :class="$style.parent">
                <nobr>
                    <BButton
                        v-if="isView('planning-backlog')"
                        size="sm"
                        variant="outline-primary"
                        title="Add to sprint"
                        :class="$style.pull"
                        @click="addToSprint">

                        <VIcon name="arrow-up"/>
                        Move to sprint
                    </BButton>

                    <BButton
                        v-if="isView('planning-sprint')"
                        size="sm"
                        variant="outline-primary"
                        title="Remove from sprint"
                        :class="$style.pull"
                        @click="removeFromSprint">

                        <VIcon name="arrow-down"/>
                        Move to backlog
                    </BButton>

                    <span
                        :class="{ 'js-drag-drop': sortable, 'ml-2': true }"
                        v-if="isView(['backlog', 'planning-sprint', 'planning-backlog'])"
                        @dblclick="!sortable || startKeyboardSort()"
                        :title="sortable ? 'Drag to change order, double click to use keyboard (arrow keys to move up/down, enter to save, esc to abort)' : 'Sort by position ascending (default order) to be able to change order'">

                        <VIcon
                            name="sort"
                            label="Drag and drop to change order"
                            :class="{ [$style.fade]: !sortable }"/>

                        <strong>{{ position }}</strong>
                    </span>
                </nobr>
            </td>

            <td v-if="isView(['sprint', 'history'])">
                <strong>
                    {{ position }}
                </strong>
            </td>
            <!-- Identifier -->
            <td>
                {{ data.identifier }}
            </td>

            <!-- Story -->
            <td @click="expandedView = !expandedView">
                <VIcon class="float-right" :name="expandedView ? 'caret-up' : 'caret-down'" />

                {{ data.title }}

                <p v-show="expandedView" class="mt-2">
                    Notes: <br>
                    {{ data.description || '(no notes)' }}
                </p>
            </td>

            <td>
                {{ topic(data.topic_id) ? topic(data.topic_id).title : '(no topic)' }}
            </td>

            <!-- Story points -->
            <td>
                <nobr>
                    {{ data.points ? data.points + ' SP' : '(not&nbsp;estimated)' }}
                </nobr>
            </td>

            <!-- Status -->
            <td v-if="isView('history')">
                <span 
                    class="badge" 
                    :class="statusMap[data.status].css">
                    {{ statusMap[data.status].name }}
                </span>
            </td>

            <td v-if="isView('sprint')">
                <BDropdown 
                    variant="link" 
                    no-caret 
                    class="b-dropdown-minimal">

                    <template slot="button-content">
                        <span 
                            class="badge" 
                            :class="statusMap[data.status].css">
                            {{ statusMap[data.status].name }}
                        </span>
                    </template>

                    <BDropdownItem 
                        v-for="(status, index) in statusMap" 
                        :key="index" 
                        @click="saveStatus(index)">

                        <span 
                            class="badge" 
                            :class="[status.css]">{{ status.name }}</span>
                    </BDropdownItem>
                </BDropdown>
            </td>

            <td v-if="isView(['backlog', 'planning-backlog', 'planning-sprint'])">
                <BButton 
                    size="sm" 
                    variant="outline-primary"
                    title="Edit"
                    class="mb-1"
                    @click="startEditing">

                    <VIcon name="pencil"/>
                    Edit
                </BButton>

                <BButton 
                    size="sm" 
                    variant="outline-danger"
                    title="Delete" 
                    v-confirm="{ action: remove, text: 'Are you sure you want to delete this story?' }">

                    <VIcon name="trash"/>
                    Delete
                </BButton>
            </td>
        </template>
    </tr>
</template>

<script>
import '@icons/sort';
import '@icons/pencil';
import '@icons/trash';
import '@icons/arrow-up';
import '@icons/arrow-down';
import '@icons/caret-down';
import '@icons/caret-up';

import * as _ from 'lodash';
import VIcon from 'vue-awesome/components/Icon';
import BButton from '@bootstrap/button/button';
import BDropdown from '@bootstrap/dropdown/dropdown';
import BDropdownItem from '@bootstrap/dropdown/dropdown-item';
import StoryForm from '@/components/forms/StoryForm';
import bus from '@/helper/bus';

export default {
    name: 'StoryItem',
    components: { VIcon, StoryForm, BButton, BDropdown, BDropdownItem },
    props: {
        data: {
            type: Object,
            default: () => {},
        },
        view: {
            type: String,
            default: 'backlog',
        },
        position: {
            type: Number,
            default: null,
        },
        sortable: {
            type: Boolean,
            default: false,
        },
        highlight: {
            type: Boolean,
            default: false,
        },
    },
    data() {
        return {
            editing: false,
            editingData: null,
            statusMap: this.$store.state.stories.statusMap,
            keyboardSorting: false,
            expandedView: false,
        };
    },
    computed: {
        project() {
            return this.$store.getters['projects/byId'](this.data.project_id);
        },
        nextSprint() {
            return this.project
                ? this.$store.getters['sprints/next'](this.project.course_id)
                : null;
        }
    },
    methods: {
        /**
         * Check if we're operating in a given view
         * @param {(string|array)} view 
         * @returns {boolean}
         */
        isView(view) {
            if (Array.isArray(view)) {
                return view.indexOf(this.view) > -1;
            } else {
                return this.view === view;
            }
        },

        topic(topicId) {
            return this.$store.getters['topics/byId'](topicId);
        },

        /**
         * Start editing process of this story
         */
        startEditing() {
            this.editing = true;
            this.editingData = _.pick(this.data, [
                'title',
                'description',
                'points',
                'topic_id',
            ]);

            this.$store.commit('newPendingChange');

            bus.$on('saveAll', this.save);
        },

        cancelEdit() {
            this.editing = false;
            this.$store.commit('resolvePendingChange');
            bus.$off('saveAll', this.save);
        },

        startKeyboardSort() {
            this.keyboardSorting = true;
            window.addEventListener('keydown', this.move);
            this.$emit('moveStart', this.data.id);
        },

        stopKeyboardSort() {
            this.keyboardSorting = false;
            window.removeEventListener('keydown', this.move);
        },

        move(evt) {
            if (this.keyboardSorting) {
                switch(evt.keyCode) {
                    // enter
                    case 13:
                        evt.preventDefault();
                        this.stopKeyboardSort();
                        this.$emit('moveComplete', this.data.id);
                        break;
                    // esc
                    case 27:
                        evt.preventDefault();
                        this.stopKeyboardSort();
                        this.$emit('moveAbort', this.data.id);
                        break;
                    // arrow up
                    case 38:
                        evt.preventDefault();
                        this.$emit('move', this.data.id, -1);
                        break;
                    // arrow down
                    case 40:
                        evt.preventDefault();
                        this.$emit('move', this.data.id, 1);
                        break;
                    default:
                        // do nothing
                }
            }
        },

        /**
         * Save the edited parameters of this story
         */
        async save() {
            await this.$store.dispatch('stories/update', {
                id: this.data.id,
                parentId: this.data.project_id,
                ...this.editingData,
            });

            this.editing = false;

            this.$store.commit('resolvePendingChange');
            bus.$off('saveAll', this.save);

            // TODO handle errors in UI
        },

        /**
         * Delete this story
         */
        async remove() {
            await this.$store.dispatch('stories/remove', {
                id: this.data.id,
                parentId: this.data.project_id,
            });
        },

        /**
         * Save new status of story
         * @param {int} statusId
         */
        async saveStatus(statusId) {
            await this.$store.dispatch('stories/update', {
                id: this.data.id,
                parentId: this.data.project_id,
                status: statusId,
            });

            // TODO handle errors in UI
        },

        /**
         * Add story to sprint
         */
        async addToSprint() {
            if (!this.nextSprint) {
                return;
            }

            await this.$store.dispatch('stories/update', {
                id: this.data.id,
                parentId: this.project.id,
                sprint_id: this.nextSprint.id,
            });

            await this.$store.dispatch('projects/fetch', {
                id: this.project.id,
            });

            // TODO handle errors in UI
        },

        /**
         * Remove story from sprint
         */
        async removeFromSprint() {
            await this.$store.dispatch('stories/update', {
                id: this.data.id,
                parentId: this.project.id,
                sprint_id: null,
            });

            await this.$store.dispatch('projects/fetch', {
                id: this.project.id,
            });

            // TODO handle errors in UI
        },
    },
};
</script>

<style lang="scss" module>
    .fade {
        fill: #ccc;
    }

    .highlight td {
        background: fade_out(#76b900, 0.3);
    }

    .parent {
        position: relative;
    }

    .pull {
        position: absolute;
        right: 115px;
        cursor: pointer;
    }
</style>
