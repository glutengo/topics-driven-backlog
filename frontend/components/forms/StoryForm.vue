<template>
    <BForm 
        @submit.prevent="submit" 
        class="story-form">

        <BRow>
            <BCol md="10">
                <BFormGroup 
                    label="Story *" 
                    label-for="story-title">

                    <BFormInput 
                        id="story-title" 
                        v-model="data.title" 
                        required/>
                </BFormGroup>
            </BCol>
            <BCol md="2">
                <BFormGroup 
                    label="Story points" 
                    label-for="story-points">

                    <BFormInput 
                        id="story-points" 
                        type="number" 
                        min="0" 
                        v-model="data.points"/>
                </BFormGroup>
            </BCol>
        </BRow>

        <BFormGroup 
            label="Notes" 
            label-for="story-description">

            <BFormTextarea 
                id="story-description" 
                v-model="data.description"/>
        </BFormGroup>

        <BButton 
            type="submit" 
            variant="primary" 
            :disabled="isInvalid">Save</BButton>

        <BButton 
            type="button" 
            variant="grey" 
            @click="cancel">Cancel</BButton>
    </BForm>
</template>

<script>
import { required, numeric, minValue } from 'vuelidate/lib/validators';
import BForm from '@bootstrap/form/form';
import BFormGroup from '@bootstrap/form-group/form-group';
import BFormInput from '@bootstrap/form-input/form-input';
import BFormTextarea from '@bootstrap/form-textarea/form-textarea';
import BRow from '@bootstrap/layout/row';
import BCol from '@bootstrap/layout/col';
import BButton from '@bootstrap/button/button';
import BaseForm from '@/components/forms/BaseForm';

export default {
    name: 'StoryForm',
    components: { BForm, BFormGroup, BFormInput, BFormTextarea, BRow, BCol, BButton },
    extends: BaseForm,
    validations: {
        data: {
            title: { required },
            points: { numeric, minValue: minValue(0) },
        },
    },
};
</script>