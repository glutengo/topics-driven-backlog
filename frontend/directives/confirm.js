import Vue from 'vue';

/**
 * v-confirm directive, asking the user to confirm the action before
 * a click event on an item is processed
 */
Vue.directive('confirm', {
    bind(el, binding) {
        el.handleClick = () => {
            if (confirm('Are you sure? This action is irreversible.')) {
                binding.value();
            }
        };

        el.addEventListener('click', el.handleClick);
    },
    unbind(el) {
        el.removeEventListener('click', el.handleClick);
    },
});
