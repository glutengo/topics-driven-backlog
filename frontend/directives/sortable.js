import Vue from 'vue';
import * as Sortable from 'sortablejs/Sortable';

/**
 * v-sortable directive, binding Sortable plugin to a list
 */
Vue.directive('sortable', {
    inserted: function(el, binding, vnode) {
        if (binding.value) {
            vnode.context.sortableInstance = Sortable.create(el, binding.value);
        }
    },
});
