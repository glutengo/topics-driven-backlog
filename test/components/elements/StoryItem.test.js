import { mount, shallow, createLocalVue } from 'vue-test-utils';
import Vuex from 'vuex';
import Vuelidate from 'vuelidate';
import StoryItem from '@/components/elements/StoryItem';
import StoryForm from '@/components/forms/StoryForm';
import '@/directives/confirm';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Vuelidate);

describe('StoryItem.test.js', () => {
    let cmp, simpleCmp, actions, store;

    beforeEach(() => {
        actions = {
            patch: jest.fn(),
            delete: jest.fn(),
        };

        store = new Vuex.Store({
            modules: {
                stories: {
                    namespaced: true,
                    state: {
                        statusMap: [{ name: 'open', css: 'badge-dark' }],
                    },
                    actions,
                },
            },
        });

        cmp = mount(StoryItem, {
            localVue,
            store,
            propsData: {
                data: {
                    identifier: 'S-001',
                    title: 'Test story',
                    points: 5,
                    status: 0,
                },
                view: 'backlog',
            },
        });

        simpleCmp = shallow(StoryItem, {
            localVue,
            store,
            propsData: {
                data: {
                    identifier: 'S-001',
                    title: 'Test story',
                    points: 5,
                    status: 0,
                },
                view: 'backlog',
            },
        });
    });

    it('tells us if we are operating in a given view', () => {
        expect(cmp.vm.isView('backlog')).toBe(true);
        expect(cmp.vm.isView('something-else')).toBe(false);
        expect(cmp.vm.isView(['backlog', 'something-else'])).toBe(true);
    });

    it('calls starts editing when edit button is clicked', () => {
        cmp.vm.startEditing = jest.fn();
        cmp.update();

        cmp
            .findAll('.btn')
            .at(0)
            .trigger('click');

        expect(cmp.vm.startEditing).toBeCalled();
    });

    it('is in edit mode after startEditing was called', () => {
        cmp.vm.startEditing();
        cmp.update();
        expect(cmp.vm.editing).toBe(true);
        expect(cmp.vm.editingData).toBeTruthy();
        expect(cmp.element).toMatchSnapshot();
    });

    it('triggers story save when story form triggers submit', () => {
        cmp.vm.startEditing();
        cmp.vm.save = jest.fn();
        cmp.update();

        cmp.find(StoryForm).trigger('submit');
        expect(cmp.vm.save).toBeCalled();
    });

    it('patches the story and leaves editing mode on story save', () => {
        cmp.vm.save();

        expect(actions.patch).toBeCalled();
        expect(cmp.vm.editing).toBe(false);
    });

    it('patches the story status correctly', () => {
        cmp.vm.saveStatus(1);

        expect(actions.patch).toBeCalled();
        expect(cmp.vm.editing).toBe(false);
    });

    it('deletes the sprint when delete button is clicked', () => {
        const confirmSpy = jest
            .spyOn(window, 'confirm')
            .mockImplementation(() => true);

        cmp
            .findAll('.btn')
            .at(1)
            .trigger('click');

        expect(confirmSpy).toHaveBeenCalled();
        expect(actions.delete).toHaveBeenCalled();
    });

    it('emits addToSprint event to parent', () => {
        const props = cmp.props;
        props.view = 'planning-backlog';
        cmp.setProps(props);

        const stub = jest.fn();
        cmp.vm.$on('addToSprint', stub);

        cmp
            .findAll('.btn')
            .at(1)
            .trigger('click');

        expect(stub).toBeCalled();
    });

    it('emits removeFromSprint event to parent', () => {
        const props = cmp.props;
        props.view = 'planning-sprint';
        cmp.setProps(props);

        const stub = jest.fn();
        cmp.vm.$on('removeFromSprint', stub);

        cmp
            .findAll('.btn')
            .at(0)
            .trigger('click');

        expect(stub).toBeCalled();
    });

    it('shows backlog page', () => {
        expect(simpleCmp.findAll('td').length).toBe(5);
        expect(simpleCmp.element).toMatchSnapshot();
    });

    it('shows history page', () => {
        const props = simpleCmp.props;
        props.view = 'history';
        simpleCmp.setProps(props);

        expect(simpleCmp.findAll('td').length).toBe(4);
        expect(simpleCmp.element).toMatchSnapshot();
    });

    it('shows planning-sprint page', () => {
        const props = simpleCmp.props;
        props.view = 'planning-sprint';
        simpleCmp.setProps(props);

        expect(simpleCmp.findAll('td').length).toBe(5);
        expect(simpleCmp.element).toMatchSnapshot();
    });

    it('shows planning-backlog page', () => {
        const props = simpleCmp.props;
        props.view = 'planning-backlog';
        simpleCmp.setProps(props);

        expect(simpleCmp.findAll('td').length).toBe(4);
        expect(simpleCmp.element).toMatchSnapshot();
    });

    it('shows sprint page', () => {
        const props = simpleCmp.props;
        props.view = 'sprint';
        simpleCmp.setProps(props);

        expect(simpleCmp.findAll('td').length).toBe(4);
        expect(simpleCmp.element).toMatchSnapshot();
    });
});
