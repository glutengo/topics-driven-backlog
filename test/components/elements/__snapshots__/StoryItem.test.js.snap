// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryItem.test.js is in edit mode after startEditing was called 1`] = `
<tr
  class="story-item"
>
  <td
    colspan="5"
  >
    <form
      class="story-form"
    >
      <div
        class="row"
      >
        <div
          class="col-md-10"
        >
          <div
            aria-labelledby="__BVID__39___BV_label_"
            class="b-form-group form-group"
            id="__BVID__39_"
            role="group"
          >
            <label
              class="col-form-label pt-0"
              for="story-title"
              id="__BVID__39___BV_label_"
            >
              Story *
            </label>
            <div
              class=""
            >
              <input
                aria-required="true"
                class="form-control"
                id="story-title"
                required="required"
                type="text"
              />
              <!---->
              <!---->
              <!---->
            </div>
          </div>
        </div>
         
        <div
          class="col-md-2"
        >
          <div
            aria-labelledby="__BVID__41___BV_label_"
            class="b-form-group form-group"
            id="__BVID__41_"
            role="group"
          >
            <label
              class="col-form-label pt-0"
              for="story-points"
              id="__BVID__41___BV_label_"
            >
              Story points
            </label>
            <div
              class=""
            >
              <input
                class="form-control"
                id="story-points"
                min="0"
                type="number"
              />
              <!---->
              <!---->
              <!---->
            </div>
          </div>
        </div>
      </div>
       
      <div
        aria-labelledby="__BVID__43___BV_label_"
        class="b-form-group form-group"
        id="__BVID__43_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="story-description"
          id="__BVID__43___BV_label_"
        >
          Notes
        </label>
        <div
          class=""
        >
          <textarea
            class="form-control"
            id="story-description"
            rows="1"
            wrap="soft"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
       
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save
      </button>
       
      <button
        class="btn btn-grey"
        type="button"
      >
        Cancel
      </button>
    </form>
  </td>
</tr>
`;

exports[`StoryItem.test.js shows backlog page 1`] = `
<tr
  class="story-item"
>
  <td>
    <span
      class="js-drag-drop"
    >
      <!---->
    </span>
  </td>
   
  <td>
    
            S-001
        
  </td>
   
  <td>
    
            Test story
        
  </td>
   
  <td>
    
            5 SP
        
  </td>
   
  <!---->
   
  <!---->
   
  <td>
    <!---->
     
    <!---->
  </td>
   
  <!---->
   
  <!---->
</tr>
`;

exports[`StoryItem.test.js shows history page 1`] = `
<tr
  class="story-item"
>
  <!---->
   
  <td>
    
            S-001
        
  </td>
   
  <td>
    
            Test story
        
  </td>
   
  <td>
    
            5 SP
        
  </td>
   
  <td>
    <span
      class="badge badge-dark"
    >
      
                open
            
    </span>
  </td>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
</tr>
`;

exports[`StoryItem.test.js shows planning-backlog page 1`] = `
<tr
  class="story-item"
>
  <!---->
   
  <td>
    
            S-001
        
  </td>
   
  <td>
    
            Test story
        
  </td>
   
  <td>
    
            5 SP
        
  </td>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <td>
    <!---->
     
    <!---->
  </td>
   
  <!---->
</tr>
`;

exports[`StoryItem.test.js shows planning-sprint page 1`] = `
<tr
  class="story-item"
>
  <td>
    <span
      class="js-drag-drop"
    >
      <!---->
    </span>
  </td>
   
  <td>
    
            S-001
        
  </td>
   
  <td>
    
            Test story
        
  </td>
   
  <td>
    
            5 SP
        
  </td>
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <td>
    <!---->
  </td>
</tr>
`;

exports[`StoryItem.test.js shows sprint page 1`] = `
<tr
  class="story-item"
>
  <!---->
   
  <td>
    
            S-001
        
  </td>
   
  <td>
    
            Test story
        
  </td>
   
  <td>
    
            5 SP
        
  </td>
   
  <!---->
   
  <td>
    <!---->
  </td>
   
  <!---->
   
  <!---->
   
  <!---->
</tr>
`;
